BidCat/API/Auction.cs

```diff
@@ -9,40 +9,179 @@ namespace BidCat.API
 {
 	public class Auction
 	{
-		Dictionary<string, Dictionary<int, int>> lotDict = new Dictionary<string, Dictionary<int, int>>();
+		// Lot ID, User ID, Amount
+		private Dictionary<string, List<Tuple<int, int>>> lotDict = new Dictionary<string, List<Tuple<int, int>>>();
 
-		AbstractBank bank;
-		public Auction(Action<ApiLogMessage> logger)
-		{
-			switch (Listener.Config.StorageType)
-			{
-				case StorageType.Memory:
-					bank = new MemoryBank(logger);
-					break;
-				case StorageType.Mongo:
-					bank = new MongoBank(logger);
-					break;
-				case StorageType.Postgres:
-					bank = new PostgresBank(logger);
-					break;
-			}
+		private List<string> ChangesTracker = new List<string>();
 
+		AbstractBank bank;
+		public Auction(Action<ApiLogMessage> logger, AbstractBank bank)
+		{
+			this.bank = bank;
+			this.bank.RegisterReservedMoneyChecker(GetReservedMoney);
 		}
 
-		public async Task<int> GetReservedMoney(int userId)
+		private async Task<int> GetReservedMoney(int userId)
 		{
 			return (await GetBidsForUser(userId)).Sum(x => x.Value);
 		}
 
+		public async Task<Dictionary<int, int>> GetReservedMoney(IEnumerable<int> userIds)
+		{
+			Dictionary<int, int> ret = new Dictionary<int, int>();
+			foreach (int userId in userIds)
+				ret.Add(userId, await GetReservedMoney(userId));
+			return ret;
+		}
+
+#pragma warning disable 1998
 		public async Task<Dictionary<string, int>> GetBidsForUser(int userId)
 		{
 			Dictionary<string, int> bids = new Dictionary<string, int>();
-			foreach (KeyValuePair<string, Dictionary<int, int>> pair in lotDict)
+			foreach (KeyValuePair<string, List<Tuple<int, int>>> pair in lotDict)
 			{
-				if (pair.Value.ContainsKey(userId))
-					bids.Add(pair.Key, pair.Value[userId]);
+				if (pair.Value.Any(x => x.Item1 == userId))
+					bids.Add(pair.Key, pair.Value.First(x => x.Item1 == userId).Item2);
 			}
 			return bids;
 		}
+#pragma warning restore 1998
```

Comment:
Again, as I said in MusicCat, please don't disable this warning. This warning exists to remind you that async isn't doing anything and you need to rewrite the function if you don't want it to block.